### DISCUSSION TITLE
Help with the references in my dsl

### DISCUSSION THREAD

**Discussion by Anthie123** [eclipse-langium/langium#1194] [Q&A] [answered]
Hi,
I started with langium for the first time and I tried to create a short Testdsl: 
```
grammar TestDsl

entry Model:
    (
    cases+=Case 
    | tests+=Test
    )*;

Case:
    'Case' name=ID url=STRING
        (functions+=FunctionDefinition)*
    'endCase';

FunctionDefinition:
    'fun' name=ID '(' (parameters+=ParameterDeclaration (',' parameters=ParameterDeclaration)* )?')' ;

ParameterDeclaration:
    name=ID ':' type=Type;

Type:
    type='String' | 'Number';

Test:
    'Test' test=[Case:ID]
        (testfunctions+=Testfunction)*
    'endTest';

Testfunction:
    tests=[FunctionDefinition:ID] ('(' (parameters+=Parameter (',' parameters=Parameter)* )? ')')? ;

Parameter:
    parameter=STRING | NUMBER;

terminal ID: /[_a-zA-Z][\w_]*/;
terminal STRING: /"[^"]*"|'[^']*'/;
terminal NUMBER: /\d+/;

hidden terminal WS: /\s+/;
hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```
and here a short example in this language: 
```
Case MyTestName "path/to/page"
    fun login(login: String, password: String)
    fun process()
endCase

Test MyTestName
    login("tets", "test")
    proceed
endTest
```
My problem is that I get errors when I try to use the "login" and "proceed" definition of the "Case" in "Test". My assumption is that I have to override the DefaultScopeProvider because the reference of "test" in "Test" have to be present when  the Editor tries to resolve the references of the "Testfunction". 

I tried to solve that overriding the ScopeProvicer in the Module of my DSL.

In my ScopeProvider I tried this:
```
    override getScope(context: ReferenceInfo): Scope {
        if (context.property === 'tests' && isTest(context.container)) {
            const testCase: Case | undefined = context.reference.ref as Case
            if (testCase) {
                return this.createScopeForNodes(testCase.functions)
            }
        }
        return super.getScope(context)
    }
```
But I am not sure if this code is right and will do what it should.
Could anybody give me a little hint how to solve my problem with the missing references?

Best regards,
Andrej

**Comment by msujew** [comment, marked as answer]
Hey @Anthie123,

the scope provider snippet that you provided isn't quite correct. The `context.container` property contains the owner of the reference. Since the `tests` reference is on the `Testfunction` type, the `isTest(context.container)` check will never return true. I haven't tested it, but something like this should work in that case:

```ts
    override getScope(context: ReferenceInfo): Scope {
        if (context.property === 'tests' && isTestfunction(context.container)) {
            const testFunction = context.container;
            // The container of the `Testfunction` type should be a `Test` element
            const testFunctionContainer = testFunction.$container as Test;
            // Now we can find the referenced `Case` element
            const testCase = testFunctionContainer.test.ref;
            if (testCase) {
                return this.createScopeForNodes(testCase.functions)
            } else {
                // Ensure that we don't return the default scope
                return EMPTY_SCOPE;
            }
        }
        return super.getScope(context)
    }
```

Note that you're not allowed to use `context.reference.ref` in the scope provider as that will throw you into an infinite loop. The scope provider is there to find possible elements that the reference can resolve to. Calling `ref` will try to execute the reference resolving algorithm - which lands us directly into the scope provider again.

The `context.reference` property is there to get info on the reference like its document position or the text that was parsed in the document for the dedicated reference. It's usually not useful to access that information during the scope provider though.

And yes, that is generally the way to write those domain specific scoping rules. We also have a few guides on that topic [here](https://langium.org/guides/scoping/) if you haven't looked into those already.

**Reply by Anthie123** [reply]
Hi Mark,
thank you for your answer. My code will work now. 

The guides I have already read but I think I will get the book you mentioned in the class member scoping chapter and get more basic knowledge about creating DSLs

Best regards,
Andrej

**Reply by msujew** [reply]
While Crafting Interpreters is a good read to learn about basic and more advanced topics in interpreter technology, I would recommend to read Markus Völter's [DSL Engineering](http://voelter.de/data/books/markusvoelter-dslengineering-1.0.pdf) if you really want to go in depth on language design and compiler frontends (Langium is effectively a compiler frontend; interpreters are compiler backends).

**Reply by Anthie123** [reply]
Thank you for the link, I will give it a try.

**Reply by snarkipus** [reply]
I also found [Implementing Domain-Specific Languages with Xtext and Xtend](https://www.amazon.com/Implementing-Domain-Specific-Languages-Xtext/dp/1786464969) incredibly helpful. Despite being based on Xtext/Xtend, the accompanying Github repo and treatment of TDD in a DSL development workflow translate very well to Langium.

**Reply by Anthie123** [reply]
Thanks for the link. I have already found this book but I did not like the focus on Xtext/Xtend.
